cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -259,72 +271,138 @@ void accumulate_vertex_results(
+  edge_src_property_t<vertex_t, vertex_t> src_distances(handle, graph_view);
+  edge_src_property_t<vertex_t, edge_t> src_sigmas(handle, graph_view);
+  edge_src_property_t<vertex_t, weight_t> src_deltas(handle, graph_view);
+  edge_dst_property_t<vertex_t, vertex_t> dst_distances(handle, graph_view);
+  edge_dst_property_t<vertex_t, edge_t> dst_sigmas(handle, graph_view);
+  edge_dst_property_t<vertex_t, weight_t> dst_deltas(handle, graph_view);
 
+  // Update all 3 properties initially (deltas start as 0)
   update_edge_src_property(
     handle,
     graph_view,
     thrust::make_zip_iterator(distances.begin(), sigmas.begin(), deltas.begin()),
-    src_properties.mutable_view());
+    view_concat(src_distances.mutable_view(), src_sigmas.mutable_view(), src_deltas.mutable_view()));
   update_edge_dst_property(
     handle,
     graph_view,
     thrust::make_zip_iterator(distances.begin(), sigmas.begin(), deltas.begin()),
-    dst_properties.mutable_view());
+    view_concat(dst_distances.mutable_view(), dst_sigmas.mutable_view(), dst_deltas.mutable_view()));
+
+  // Use binary search method to find frontier boundaries more efficiently
+  std::vector<vertex_t> h_bounds{};
+  rmm::device_uvector<vertex_t> vertices_sorted(distances.size(), handle.get_stream());
+  {
+    // Create distance_keys for sorting
+    rmm::device_uvector<vertex_t> distance_keys(distances.size(), handle.get_stream());
+    
+    // Copy distances for sorting (preserve original)
+    raft::copy(distance_keys.data(), distances.data(), distances.size(), handle.get_stream());
+
+    // Use thrust::sequence instead of thrust::copy for vertices
+    thrust::sequence(handle.get_thrust_policy(),
+                    vertices_sorted.begin(), vertices_sorted.end(),
+                    graph_view.local_vertex_partition_range_first());
+    
+    // Sort vertices by distance using stable_sort_by_key
+    thrust::stable_sort_by_key(
+      handle.get_thrust_policy(),
+      distance_keys.begin(), distance_keys.end(),   // keys (copied distances)
+      vertices_sorted.begin()                       // values (vertices)
+    );
+
+    rmm::device_uvector<vertex_t> d_bounds(diameter + 1, handle.get_stream());
+    
+    // Single vectorized thrust call to compute all bounds for distances 0 to diameter
+    thrust::lower_bound(
+      handle.get_thrust_policy(),
+      distance_keys.begin(), distance_keys.end(),   // sorted distances
+      thrust::make_counting_iterator<vertex_t>(0),  // search keys: [0, 1, 2, 3, ...]
+      thrust::make_counting_iterator<vertex_t>(diameter + 1),
+      d_bounds.data());
+    
+    // Copy bounds to host for use in delta loop
+    h_bounds.resize(d_bounds.size());
+    raft::update_host(h_bounds.data(), d_bounds.data(), d_bounds.size(), handle.get_stream());
+    handle.sync_stream();
+  }
+  
+  // Calculate max frontier size using the precomputed bounds
+  vertex_t max_frontier_size = 0;
+  for (vertex_t d = 0; d < h_bounds.size() - 1; ++d) {
```

Comment:
Please don't ignore compiler warnings.

You will see "comparison of integer expressions of different signedness". Fix it.